Bryanv/categoricals

bryevdv opened the pull request:
issues: closes #4369
issue: fixes #516

Allows for nested categoricals, upt to three levels, much easier to produce e.g. bar charts like this:

<img width="606" alt="screen shot 2017-08-03 at 11 03 50" src="https://user-images.githubusercontent.com/1078448/28931394-8913e16c-783b-11e7-864d-73b500c3afac.png">


Some things were removed, at least for now:

* `offset` 

This was only for MPL compat which is being dropped, so I have no qualms removing it entirely. Some things were changed (these now accept synthetic coordinates, which is much simpler/more straighforward)

* `bounds`
* `min_interval` 
* `max_interval`

Planned work:

* [x] expand `factors` property to accept nested categories in addition to simple lists of factors
* [x] update `synthetic` to properly compute synthetic coords for hierarchical categories
* [x] add support for display padding between categories 
* [x] update axis to be able to display ranges for multi-level categories
* [x] make CategoricalColorMapper support nested categories
* [x] fix up JS API
* [x] computed expressions
* [x] sugar for dodging, jitter, and stacking things
* [x] add tests
* [ ] add examples
* [ ] add docs

ping @bokeh/dev

bryevdv commented:
@mattpap JS tests are failing due to one thing, I had to add this to `set_data`:
```
    # if we have any coordinates that are categorical, convert them to
    # synthetic coords here
    xr = @renderer.plot_view.frame.x_ranges[@model.x_range_name]
    yr = @renderer.plot_view.frame.y_ranges[@model.y_range_name]
    for [xname, yname] in @model._coords
      xname = "_#{xname}"
      yname = "_#{yname}"
      if xr.v_synthetic?
        @[xname] = xr.v_synthetic(@[xname])
      if yr.v_synthetic?
        @[yname] = yr.v_synthetic(@[yname])
``` 
but `x_ranges` etc are undefined, and things fall over. Any suggestions?

bryevdv commented:
Added 
```
if @renderer.plot_view.frame.x_ranges? 
``` 
check as a temp workaround for now

mattpap commented:
> but x_ranges etc are undefined, and things fall over. Any suggestions?

Are offending tests using fully initialized models? Looking at the code, if models are properly initialized, this shouldn't happen. Obviously it can be a dozen of other things. If I knew what tests are failing, I could say more.

bryevdv commented:
> Are offending tests using fully initialized models?

Sorry, it's the defaults test

canavandl commented:
I just had an idea of something to check before merging:

In the case of a VBar with the "x" values in categorical space, presumably a numerical width value should still work.

bryevdv commented:
Early example (ping @pzwang)
```
from bokeh.io import output_file, show
from bokeh.models import FactorRange
from bokeh.plotting import figure
from bokeh.sampledata.autompg import autompg as df

df.origin = df.origin.astype(str)
df.cyl = df.cyl.astype(str)
group = df.groupby(('origin', 'cyl'))
stats = group.mean()

p = figure(x_range=FactorRange(*g.groups.keys()), y_range=(0,35))
p.vbar(x=stats.index, top=stats.mpg, width=0.9, color=color)

output_file("foo.html")

show(p)
```

<img width="624" alt="screen shot 2017-07-23 at 15 11 00" src="https://user-images.githubusercontent.com/1078448/28502508-59f0f0a8-6fb9-11e7-8cbf-58b2f56a4a25.png">

@canavandl it does, when I've got a bit more done I will write up the tradeoffs. One of them is that factors have to be strings. I don't see an alternative, otherwise numbers become ambiguous "is this a factor? or an offset?"

canavandl commented:
> @canavandl it does, when I've got a bit more done I will write up the tradeoffs. One of them is that factors have to be strings. I don't see an alternative, otherwise numbers become ambiguous "is this a factor? or an offset?"

Great! I was just had the idea while making an unrelated example and wanted to ensure that edge case is handled.

bryevdv commented:
> Great! I was just had the idea while making an unrelated example and wanted to ensure that edge case is handled.

This is actually other change. Before, every category offset was a percentage, with 0 being the prev cat, and 1 being the next cat, and 0.5 being the cat in question. Now its not like that. The offset is simply a +/- added to a factor's synthetic coordinate. It's mostly all better and simpler but there is one trade off. If you have a top level category with three sub categories and you wanted a big bar 90% wide of the full top-level cat, you have to know to make the width 3*0.9 = 2.7 sucks a bit but is not a common case and also not so terrible.

mattpap commented:
> Sorry, it's the defaults test.

That makes sense, because those tests work with insufficiently configured models. However, this piece of code shouldn't be triggered by those tests.

bryevdv commented:
@mattpap I'd like to extend this to handle more levels of nesting (at least one or two more) It would really make things simple if there was a Map object that could use list values as keys. Do you know of one we could add and use that is lightweight?

mattpap commented:
@bryevdv, what operations on that map you need to perform? You could use ES6 Map, but that is very annoying to iterate over in ES5.

bryevdv commented:
@mattpap I just want to be able to do e.g.
```
mapping[ ["foo", "bar"] ] = 10
``` 
and have it be robust (i.e. not "work" by converting the list to a string and hashing that)

mattpap commented:
@bryevdv, so use ES6 Map, e.g.
```
map = new Map()
map.set([1, 2, 3], ["a", "b", "c"])
map.get([1, 2, 3])
map.has([1, 2, 3])
```
If you need iteration then there is `forEach`:
```
map.forEach((key, value) -> ...)
```

bryevdv commented:
@mattpap the `.has` method seems to be failing, at least with lists:

<img width="369" alt="screen shot 2017-07-25 at 17 00 06" src="https://user-images.githubusercontent.com/1078448/28595730-bf06e4f0-715a-11e7-969d-b5994d59961e.png">

mattpap commented:
@bryevdv, right, it uses `===` (with a special case for `NaN`) semantics for comparison, and `===` is useless for pretty much anything than primitives.

I will implement something more reasonable, as I need a proper map for my own work anyway.

canavandl commented on bokehjs/src/coffee/models/ranges/factor_range.coffee:
```diff
@@ -1,62 +1,167 @@
+    factor_padding:      [ p.Number,       0         ]
+    subgroup_padding:    [ p.Number,       0.8       ]
+    group_padding:       [ p.Number,       1.4       ]
+    factor_padding:      [ p.Number,       0         ]
```
factor_padding is here twice

bryevdv commented:
@mattpap @canavandl @clairetang6 I still need to add docs and examples, but this is basically ready for at least a preliminary technical review.

mattpap commented on bokehjs/src/coffee/models/mappers/categorical_color_mapper.coffee:
```diff
@@ -1,22 +1,56 @@
+      return false
+  return true
+
+_find = (a, b) ->
```
`core/util/array` has `findIndex()`.

bryevdv replied:
Thanks @mattpap I have updated and added full tests for the categorical colormapper as well

bryevdv commented:
@bokeh/dev this is marked ready, the only additional work planned are docs, examples and (possibly) a tiny convenience to accept pandas groupby's directly for factors to `figure` (very simple to do)

bryevdv commented:
@mattpap @canavandl ideally I'd like to get this merged by Fri/Sat and in a new dev build by Monday so folks can start banging on it

bryevdv commented:
I added a an optional `range` prop to `Jitter` so that the transform can work with categorical data:
```
import pandas as pd

from bokeh.io import output_file, show
from bokeh.models import ColumnDataSource, Jitter
from bokeh.plotting import figure

output_file("stats.html")

DAYS = ['Sun', 'Sat', 'Fri', 'Thu', 'Wed', 'Tue', 'Mon']

data = pd.read_csv("commits.txt", parse_dates=True, header=None, names=['day', 'datetime'], index_col='datetime')
data = data.tz_localize('GMT').tz_convert('US/Central')
data['time'] = data.index.time

source = ColumnDataSource(data)

p = figure(plot_width=800, plot_height=300, y_range=DAYS, x_axis_type='datetime', title="Bokeh Commits (US/Central) ")
p.circle(x='time', y={'field': 'day', 'transform': Jitter(width=0.7, range=p.y_range)},  source=source, alpha=0.5)

p.xaxis[0].formatter.days = ['%Hh']
p.ygrid.grid_line_color = None

show(p)
```
<img width="973" alt="screen shot 2017-08-03 at 14 18 16" src="https://user-images.githubusercontent.com/1078448/28939380-b99dcca6-7856-11e7-9fca-743925c48413.png">

jbednar commented:
Nice!

bryevdv commented:
OK, last small feature, can pass pandas groups directly to `figure` to set ranges:
```
group = df.groupby(('origin', 'cyl'))

p = figure(plot_width=1100, x_range=group, tools="hover,pan,wheel_zoom,reset")
```

bryevdv commented:
ok, really, truly last feature, CDS can now be inited with Pandas groupbys, which will create columns  from `group.describe`:
```
group = df.groupby(('origin', 'cyl'))
stats = group.mean()

source = ColumnDataSource(group)
cmap = CategoricalColorMapper(palette=Spectral5, factors=sorted(df.cyl.unique()), start=1)

p = figure(plot_width=1100, x_range=group, tools="hover,pan,wheel_zoom,reset")
p.vbar(x='origin_cyl', top='mpg_mean', width=1, line_color="white", fill_color={'field': 'origin_cyl', 'transform': cmap}, source=source, legend="origin_cyl")
```

All this goes a huge way to making `bokeh.plotting` much more friendly to categorical data and pandas as well. I will work on docs/examples tomorrow

bryevdv commented:
There was a small issue with cat color mapper. I had a bad test but picked it up writing examples. I will push the fix with some docs later today

bryevdv commented:
OK, I made one more somewhat substantial addition. I added support for computed expressions in https://github.com/bokeh/bokeh/pull/6654/commits/cc5ba864220070c77af463594a6ad40be5fee369 This has been mentioned on and off for a long time, and it is necessary to round out a good usr experience for categoricals. Fortunately in the current state, adding it was fairly straightforward and required little code change and no existing test changes. With this addition, it is possible to create a `Stack` expression that permits code like this:

```
source = ColumnDataSource(data={'fruits' : fruits,
                                '2016'   : [5, 3, 4, 2, 4, 6],
                                '2017'   : [3, 2, 4, 4, 5, 3]})

p = figure(x_range=fruits, plot_height=250,
    toolbar_location=None, title="Fruit Counts by Year")

p.vbar(x='fruits', top='2016', width=0.9, color="#718dbf",
       source=source, legend=value('2016'))
p.vbar(x='fruits', bottom=stack('2016'), top=stack('2016', '2017'),
       width=0.9, color="#e84d60", source=source, legend=value('2017'))
```
which easily yields the ever asked-for but otherwise far too-difficult stacked bar chart:

<img width="541" alt="screen shot 2017-08-04 at 21 58 51" src="https://user-images.githubusercontent.com/1078448/28993203-4f4001dc-7976-11e7-841a-c34db5d05043.png">

I've added this example to the howto notebook, and will start the user guide chapter based off the notebook tomorrow. 

@mattpap  https://github.com/bokeh/bokeh/pull/6654/commits/cc5ba864220070c77af463594a6ad40be5fee369 has good tests but please do review specifically.

bryevdv commented:
@bokeh/dev plan to merge this later today and make a dev build and put out a general call for help exercising it on the mailing list, unless there is objection

bryevdv commented:
The only notable last change was https://github.com/bokeh/bokeh/pull/6654/commits/cb5f3ae712f4916f008057b19050153c11637903 which slightly simplified CDS init from data frame to handle cases where there is a `None` index name more gracefully (test added)
